sympy/sets/conditionset.py

```diff
@@ -34,6 +34,8 @@ def __new__(cls, sym, condition, base_set):
             return S.EmptySet
         if condition == S.true:
             return base_set
+        if getattr(base_set, "is_FiniteSet", False) and base_set.is_FiniteSet:
+            return base_set.intersect(condition.as_set())
```

Comment:
I didn't realize this calls `solve`. That seems like potentially a lot of computation for a constructor. That sort of thing should rather go in the `_eval_simplify` method than `__new__`.